src/cargo/core/dependency.rs

```diff
@@ -103,6 +105,25 @@ impl DependencyInner {
         })
     }
 
+    fn parse_with_deprecated(req: &str, config: &Config) -> CargoResult<VersionReq> {
+        match VersionReq::parse(req) {
+            Err(e) => {
+                match e {
+                    ReqParseError::DeprecatedVersionRequirement(requirement) => {
+                        let msg = format!("One of your version requirements ({}) is invalid. \
+Previous versions of Cargo accepted this malformed requirement, but it is being deprecated. Please \
+use {} instead.", req, requirement);
```

Comment:
It's.... not very simple. Saying "Hey the dependency "foo" is, but not which crate has the "foo" dependency. Ugh.